src/AbstractProvider.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+/**
+ * Part of the Joomla Framework AI Package
+ *
+ */
+
+namespace Joomla\AI;
+
+use Joomla\Http\HttpFactory;
+
+/**
+ * Abstract provider class.
+ *
+ * @since  __DEPLOY_VERSION__
+ */
+abstract class AbstractProvider implements ProviderInterface
+{
+    /**
+     * The provider options.
+     *
+     * @var    array|\ArrayAccess
+     */
+    protected $options;
+
+    /**
+     * The HTTP factory instance.
+     *
+     * @var    HttpFactory
+     */
+    protected $httpFactory;
+
+    /**
+     * Constructor.
+     *
+     * @param   array|\ArrayAccess  $options  Provider options array.
+     * @param   HttpFactory   $httpFactory  The http factory
+     * 
+     * @throws  \InvalidArgumentException  If options is not an array or does not implement ArrayAccess.
+     * @since  ___DEPLOY_VERSION___
+     */
+    public function __construct($options = [], ?HttpFactory $httpFactory = null)
+    {
+        // To do: Exception Handeling Code
+        // Validate provider is suported
+        // Validate that $options is an array or implements ArrayAccess
+        if (!\is_array($options) && !($options instanceof \ArrayAccess)) {
+            throw new \InvalidArgumentException(
+                'The options param must be an array or implement the ArrayAccess interface.'
+            );
+        }
+
+        $this->options = $options;
+        $this->httpFactory = $httpFactory ?: new HttpFactory();
```

Comment:
```HttpFactory::getHttp()``` is a static method, so need to check if calling it in a object context raises any notice/errors.